File: 16_bit_manipulation/02_counting_bits/counting_bits.py
from typing import List


class Solution:
    def count_bits_in_num(self, n: int) -> int:
        bit_count: int = 0
        for i in range(32):
            if n & (1 << i):
                bit_count += 1
            if (1 << i) > n:
                continue
        return bit_count

    def countBits(self, n: int) -> List[int]:
        bits = [self.count_bits_in_num(num) for num in range(n + 1)]
        return bits

File: 16_bit_manipulation/02_counting_bits/counting_bits.cpp
#include <vector>

using namespace std;

class Solution
{
  public:
    vector<int> countBits(int n)
    {
        vector<int> bits(n + 1);
        for (int i = 0; i < n + 1; ++i)
            bits[i] = count_bits_in_num(i);
        return bits;
    }
  private:
    int count_bits_in_num(int num)
    {
        int bits_count = 0;
        for (int i = 0; i < 32; ++i)
        {
            if (num & (1 << i))
                bits_count++;
            if ((1 << i) > num)
                return bits_count;
        }
        return bits_count;
    }
};